kaspad/src/daemon.rs

```diff
@@ -308,13 +309,92 @@ do you confirm? (answer y/n or pass --yes to the Kaspad command line to confirm
         && (meta_db.get_pinned(b"multi-consensus-metadata-key").is_ok_and(|r| r.is_some())
             || MultiConsensusManagementStore::new(meta_db.clone()).should_upgrade().unwrap())
     {
-        let msg =
-            "Node database is from a different Kaspad *DB* version and needs to be fully deleted, do you confirm the delete? (y/n)";
-        get_user_approval_or_exit(msg, args.yes);
+        let mut mcms = MultiConsensusManagementStore::new(meta_db.clone());
+        let version = mcms.version().unwrap();
+
+        // TODO: Update this entire section to a more robust implementation that allows applying multiple upgrade strategies.
+        // If I'm at version 3 and current version is 7, I need to be able to upgrade to that version following the intermediate
+        // steps without having to delete the DB
+        if version == 3 {
+            let active_consensus_dir_name = mcms.active_consensus_dir_name().unwrap();
+
+            match active_consensus_dir_name {
+                Some(current_consensus_db) => {
+                    // Apply soft upgrade logic: delete GD data from higher levels
+                    // and then update DB version to 4
+                    let consensus_db = kaspa_database::prelude::ConnBuilder::default()
+                        .with_db_path(consensus_db_dir.clone().join(current_consensus_db))
+                        .with_files_limit(1)
+                        .build()
+                        .unwrap();
+                    info!("Scanning for deprecated records to cleanup");
+
+                    let mut batch = rocksdb::WriteBatch::default();
+                    let mut gd_record_count: u32 = 0;
+                    let mut compact_record_count: u32 = 0;
+
+                    for level in 1..(config.max_block_level + 1) {
+                        let level_bytes = level.to_le_bytes();
+                        let ghostdag_prefix_vec = DatabaseStorePrefixes::Ghostdag.into_iter().chain(level_bytes).collect_vec();
+                        let ghostdag_prefix = ghostdag_prefix_vec.as_slice();
+
+                        for result in consensus_db.iterator(rocksdb::IteratorMode::From(ghostdag_prefix, rocksdb::Direction::Forward))
```

Comment:
It is possible to delete by range which is more efficient. See, e.g., here https://github.com/kaspanet/rusty-kaspa/blob/master/database/src/writer.rs#L53

Reply:
Updated to use `delete_range`:

```
// Apply delete of range from level 1 to max (+1) for Ghostdag and GhostdagCompact:
writer.delete_range(start_ghostdag_prefix_vec.clone(), end_ghostdag_prefix_vec.clone()).unwrap();
writer.delete_range(start_compact_prefix_vec.clone(), end_compact_prefix_vec.clone()).unwrap();
```